src/confluent_kafka/src/confluent_kafka.c

```diff
@@ -1461,6 +1461,59 @@ static void throttle_cb (rd_kafka_t *rk, const char *broker_name, int32_t broker
         CallState_resume(cs);
 }
 
+static int32_t partitioner_cb (const rd_kafka_topic_t *rkt,
+			       const void *keydata,
+			       size_t keylen,
+			       int32_t partition_cnt,
+			       void *rkt_opaque,
+			       void *msg_opaque) {
+
+	PyObject *callback = rkt_opaque;
+	PyObject *result = NULL, *args = NULL;
+	PyObject *keydataobj = NULL;
+
+	/* This is only used for testing */
+	if (callback != NULL) {
+		PyObject *ks;
+		PyObject *ks8 = NULL;
+		PyObject *vs = NULL, *vs8 = NULL;
+		const char *k = NULL;
+		printf("Jing Liu callback before cfl_PyObject_Unistr\n");
+		ks = cfl_PyObject_Unistr(callback);
+		printf("Jing Liu callback after cfl_PyObject_Unistr\n");
+		k = cfl_PyUnistr_AsUTF8(ks, &ks8);
+		printf("Jing Liu partitioner_cb at callback %s\n", k);
+	}
+
+	keydataobj = Py_BuildValue("s", keydata);
+
+	args = Py_BuildValue("(y#)", keydataobj, partition_cnt);
+
+	if (!args) {
+		cfl_PyErr_Format(RD_KAFKA_RESP_ERR__FAIL,
```

Comment:
Since this callback is triggered from a librdkafka thread, and PyErr sets the exception flag, I'm not sure where that exception will bubble up.  If we don't do something smart here the exception will probably be raised by whatever Python call the script is doing, which isn't right.

I'm not sure what the best solution is. Perhaps just to print the exception and move on, alternatively to enqueue an error for the next poll() call.